hack-sound-server

```diff
@@ -0,0 +1,158 @@
+        elif message.type == Gst.MessageType.ERROR:
+            error, debug = message.parse_error()
+            self.emit("error", error, debug)
+
+class HackSoundServer(Gio.Application):
+
+    _DBUS_NAME = "com.endlessm.HackSoundServer"
+    _DBUS_UNKNOWN_SOUND_EVENT_ID = "com.endlessm.HackSoundServer.UnknownSoundEventID"
+    _DBUS_GST_ERROR = "com.endlessm.HackSoundServer.GStreamerError"
+    _DBUS_XML = """
+    <node>
+      <interface name='com.endlessm.HackSoundServer'>
+        <method name='PlaySound'>
+          <arg type='s' name='sound_event' direction='in'/>
+          <arg type='s' name='uuid' direction='out'/>
+        </method>
+        <signal name='Error'>
+          <arg type='s' name='uuid'/>
+          <arg type='s' name='error_message'/>
+          <arg type='s' name='error_domain'/>
+          <arg type='i' name='error_code'/>
+          <arg type='s' name='debug'/>
+        </signal>
+      </interface>
+    </node>
+    """
+
+    def __init__(self, metadata):
+        super().__init__(application_id=self._DBUS_NAME,
+                         flags=Gio.ApplicationFlags.IS_SERVICE)
+        self._dbus_id = None
+        self.metadata = metadata
+        self.hold()
+        self.players = {}
+
+    def do_dbus_register(self, connection, path):
+        info = Gio.DBusNodeInfo.new_for_xml(self._DBUS_XML)
+        self._dbus_id = connection.register_object(path,
+                                                   info.interfaces[0],
+                                                   self.__method_called_cb)
+        return Gio.Application.do_dbus_register(self, connection, path)
+
+    def do_dbus_unregister(self, connection, path):
+        Gio.Application.do_dbus_unregister(self, connection, path)
+        if not self._dbus_id:
+            return
+        connection.unregister_object(self._dbus_id)
+        self._dbus_id = None
+
+    def play_sound(self, sound_event_id, connection, sender, path, iface,
+                   invocation):
+        if sound_event_id not in self.metadata:
+            invocation.return_dbus_error(self._DBUS_UNKNOWN_SOUND_EVENT_ID,
+                "sound event with id %s does not exist" % sound_event_id)
+            return
+
+        uuid_ = str(uuid.uuid4())
+        self.players[uuid_] = HackSoundPlayer(self.metadata[sound_event_id],
+                                              sender)
+        self.players[uuid_].connect("eos", self.__player_eos_cb, uuid_)
+        self.players[uuid_].connect("error", self.__player_error_cb, uuid_,
+                                    connection, path, iface)
+        self.players[uuid_].play()
+        invocation.return_value(GLib.Variant('(s)', (uuid_, )))
+
+    def __method_called_cb(self, connection, sender, path, iface,
+                           method, params, invocation):
+        if method == "PlaySound":
+            self.play_sound(params[0], connection, sender, path, iface,
+                            invocation)
```

Comment:
You should end the invocation with `G_DBUS_ERROR_UNKNOWN_METHOD` if the method name isn't recognized.